An Azure AKS + Let's Encrypt Tutorial

wallrj opened the pull request:
**Preview**: https://deploy-preview-1120--cert-manager-website.netlify.app/docs/tutorials/getting-started-with-cert-manager-on-azure-kubernetes-service-using-lets-encrypt-for-ssl/

Written during the course of testing https://github.com/cert-manager/cert-manager/pull/5570
 * https://github.com/cert-manager/cert-manager/pull/5570

pinkfloydx33 commented:
AZWI can be installed as a first class feature of AKS using the `--enable-workload-identity` switch. It's behind a feature flag currently but will be the recommended way to install once the feature is GA. Not sure if it's worth including that information or not: [here](https://learn.microsoft.com/en-us/azure/aks/workload-identity-deploy-cluster)

pinkfloydx33 commented:
Also FWIW, I think most people may be _used_ to specifying the managed identity's clientID using the ServiceAccount annotation. The value on the CRD is _meant_ to act as an _override_ for cases where you are dealing with multiple DNS zones that need different MSI's for access (though it can work perfectly fine this way). 

I pushed for that feature to be added to the PR so that there was parity with pod-identity and so that cert-manager behaved similarly to other applications that have already added AZWI support. Personally, I only need interaction with one DNS zone per cluster, so specifying the clientId as a ServiceAccount annotation and omitting the `managedIdentity` object  is what I would do. I'm not sure what the common case would look like though... 

Maybe include an example of both ways? One using the annotation only and another example for "configuring cert-manager to use multiple DNS zones requiring different MSI's"?

weisdd commented on content/docs/tutorials/getting-started-with-cert-manager-on-azure-kubernetes-service-using-lets-encrypt-for-ingress-ssl/README.md:
```diff
@@ -0,0 +1,397 @@
+    --node-count 1 \
+    --node-vm-size "Standard_B2s" \
+    --load-balancer-sku basic \
+    --enable-oidc-issuer # ℹ️ This is required for workload identity federation and is only available when using the aks-preview extension.
```
Thanks for writing the detailed docs! I'll try to have a closer look at the document soon.

There's a tiny thing that I've just spotted:
OIDC issuer has been promoted to GA in AKS [2022-10-16](https://github.com/Azure/AKS/blob/master/CHANGELOG.md#release-2022-10-16), so I think the `aks-preview` extension is only required in case Workload Identity itself is installed [as an extension](https://learn.microsoft.com/en-us/azure/aks/workload-identity-deploy-cluster#register-the-enableworkloadidentitypreview-feature-flag).
```suggestion
    --enable-oidc-issuer # ℹ️ This is required for workload identity federation.
```

wallrj replied:
Done.

wallrj commented:
Thanks both of you for the feedback. I've updated the tutorial to use the --enable-workload-identity flag, 
and I've added the service account label through a `helm install --set-string` flag,
and that simplifies things quite a lot.

weisdd commented on content/docs/tutorials/getting-started-with-cert-manager-on-azure-kubernetes-service-using-lets-encrypt-for-ingress-ssl/README.md:
```diff
@@ -0,0 +1,489 @@
+    --node-vm-size "Standard_B2s" \
+    --load-balancer-sku basic \
+    --enable-oidc-issuer \
+    --enable-workload-identity # ℹ️ This is currently only available when using the aks-preview extension.
```
I think it'd be useful to mention that if `EnableWorkloadIdentityPreview` feature flag is not yet enabled for the subscription, a user will have to follow these steps: https://learn.microsoft.com/en-us/azure/aks/workload-identity-deploy-cluster#register-the-enableworkloadidentitypreview-feature-flag

wallrj commented:
/hold until 1.11 is released because this is being merged into the live docs.

wallrj commented:
> Maybe include an example of both

@pinkfloydx33 I didn't have the energy to document client-id both ways.
It seems to me that adding the client-id to the issuer spec is more appropriate because otherwise it requires a decision to be made about which managed identity to associate with the cert-manager service account by a cluster administrator at cert-manager install time.
This seems like a decision to be made by a cert-manager administrator who is responsible for configuring the clusterissuer....that's my excuse anyway :-)

pinkfloydx33 commented:
Fine by me, makes sense. Like I said, not sure what the common case will be.

weisdd commented on content/docs/tutorials/getting-started-with-cert-manager-on-azure-kubernetes-service-using-lets-encrypt-for-ssl/README.md:
```diff
@@ -0,0 +1,735 @@
+    --wait \
+    --namespace cert-manager \
+    --set installCRDs=true \
+    --set-string 'serviceAccount.labels.azure\.workload\.identity/use=true' # ❗ Label the cert-manager service account for the attention of workload-identity webhook
```
There's upcoming change in workload identity webhook, which would require the label to rather be set on pods (https://github.com/Azure/azure-workload-identity/issues/601). For the transition period, any of those would work (serviceAccount or pod label: https://github.com/Azure/azure-workload-identity/blob/df6362a239c5d3fc19f48ff8bc7ae2ced594f116/pkg/webhook/webhook.go#L136). Since we never know which version of the webhook is installed in a cluster, I guess we can make the example universal by simply setting the label in two places. 
```suggestion
    --set-string 'podLabels.azure\.workload\.identity/use=true' \
    --set-string 'serviceAccount.labels.azure\.workload\.identity/use=true' # ❗ Label the cert-manager service account and pod for the attention of workload-identity webhook
```

wallrj replied:
Done. I've moved these to a separate values.yaml file for easier readability.

weisdd commented on public/docs/tutorials/getting-started-with-cert-manager-on-azure-kubernetes-service-using-lets-encrypt-for-ssl/values.yaml:
```diff
@@ -0,0 +1,6 @@
+# values.yaml
+podLabels:
+    azure.workload.identity/use: "true"
```
```suggestion
  azure.workload.identity/use: "true"
```
The indentation is a bit off, the rest looks great :)

wallrj replied:
Done.

wallrj commented:
/hold cancel

SgtCoDFish commented on content/docs/tutorials/getting-started-with-cert-manager-on-azure-kubernetes-service-using-lets-encrypt-for-ssl/README.md:
````diff
@@ -0,0 +1,806 @@
+...
+```
+
+> 🔗 If you don't already have `envsubst` installed you can [download and install a Go implementation of `envsubst`](https://github.com/a8m/envsubst).
````
suggestion: Would this tip be more helpful if it were moved up closer to where we actually use envsubst for the first time?

SgtCoDFish commented on content/docs/tutorials/getting-started-with-cert-manager-on-azure-kubernetes-service-using-lets-encrypt-for-ssl/README.md:
````diff
@@ -0,0 +1,806 @@
+Hostname: helloweb-55cb4cd887-tjlvh
+```
+
+> ⚠️ We used curl's `--insecure` option because it rejects self-signed certificates by default.
+> Later you will learn how to create a trusted SSL certificate signed by Let's Encrypt.
````
suggestion: Do you think this info should come _before_ we use `--insecure`?

Most people should see `--insecure` as a red flag, and I think I'd want it to be crystal clear that it won't _stay_ insecure.

SgtCoDFish commented on content/docs/tutorials/getting-started-with-cert-manager-on-azure-kubernetes-service-using-lets-encrypt-for-ssl/README.md:
```diff
@@ -0,0 +1,806 @@
+
+# Part 2
+
+In part 1 you created a test SSL certificate.
+Now you will learn how to configure cert-manager to use Let's Encrypt and Azure DNS to create a trusted SSL certificate which you can use in production.
+You need to prove to Let's Encrypt that you own the domain name of the SSL certificate and one way to do this is to create a special DNS record in that domain.
```
nit (non-blocking): I'd really prefer to use "TLS certificate" and not SSL. SSL is really old at this point.

Don't think it _needs_ to be changed though!

SgtCoDFish commented on content/docs/tutorials/getting-started-with-cert-manager-on-azure-kubernetes-service-using-lets-encrypt-for-ssl/README.md:
```diff
@@ -0,0 +1,806 @@
+> we'll start with the Let's Encrypt staging issuer, and once we're happy that it's working
+> we'll switch to the production issuer.
+>
+> 📖 Read more about [configuring the ACME Issuer](../../configuration/acme).
```
suggestion: I don't think this link follows https://github.com/cert-manager/website/#task-linking-to-other-documentation-pages

SgtCoDFish commented on content/docs/tutorials/getting-started-with-cert-manager-on-azure-kubernetes-service-using-lets-encrypt-for-ssl/README.md:
````diff
@@ -0,0 +1,806 @@
+
+You should once again be able to connect to the website, but this time you will see the Let's Encrypt staging certificate:
+
+```
+$ curl -v --insecure https://www.$DOMAIN_NAME
````
suggestion: This is missing a type - it should be ```text I think.

suggestion: Is it worth pointing out (again) that we're using `--insecure` because it's LE staging, and that we'll get to a secure cert really soon?

SgtCoDFish commented on content/docs/tutorials/getting-started-with-cert-manager-on-azure-kubernetes-service-using-lets-encrypt-for-ssl/README.md:
````diff
@@ -0,0 +1,806 @@
+kubectl explain Issuer
+```
+
+> 📖 Read about [other ways to install cert-manager](../../installation).
+>
+> 📖 Read more about [Certificates and Issuers](../../concepts).
````
suggestion: again, https://github.com/cert-manager/website/#task-linking-to-other-documentation-pages

Same applies to any other internal documentation links on here; I've not got time to check them all ATM unfortunatley!

SgtCoDFish reviewed (changes requested):
Few comments; I love this overall. Gonna be a great addition - thank you for doing it!

SgtCoDFish commented on content/docs/manifest.json:
```diff
@@ -351,8 +351,12 @@
+            },
+            {
+                "title": "GKE + Ingress + Let's Encrypt",
+                "path": "/docs/tutorials/getting-started-with-cert-manager-on-google-kubernetes-engine-using-lets-encrypt-for-ingress-ssl/README.md"
```
Fair enough, makes sense to me :grin:

SgtCoDFish commented on README.md:
```diff
@@ -37,7 +37,7 @@ There are several folders in `content/` and which one you need depends on what y
   Add it to `docs/` and possibly to the specific version of cert-manager that's latest (e.g. `v1.8-docs/`)
 - Something which only applies to the next major version of cert-manager? <br />
-  Add it to `next-docs`
+  Add it to `docs/` but branch from the [`release-next` branch](https://github.com/cert-manager/website/tree/release-next) and merge the PR into that branch.
```
Great drive-by improvement :+1:

SgtCoDFish commented on content/docs/tutorials/getting-started-aks-letsencrypt/README.md:
````diff
@@ -0,0 +1,812 @@
+
+Then use `envsubst` to substitute your chosen domain name into the following Certificate template:
+
+```yaml
+# certificate.yaml
+apiVersion: cert-manager.io/v1
+kind: Certificate
+metadata:
+  name: www
+spec:
+  secretName: www-tls
+  privateKey:
+    rotationPolicy: Always
+  commonName: www.$DOMAIN_NAME
+  dnsNames:
+    - www.$DOMAIN_NAME
+  usages:
+    - digital signature
+    - key encipherment
+    - server auth
+  issuerRef:
+    name: selfsigned
+    kind: ClusterIssuer
+```
+🔗 <a href="certificate.yaml">`certificate.yaml`</a>
````
suggestion: I don't think it's required for you to do this, but seeing this just triggered a memory for me of one of the benefits of the website migration - embedding YAML files.

See for example the old ACME tutorial (which maybe should be a redirect to your newer/better one, but that's a task for another day) [here](https://github.com/cert-manager/website/blob/90ea038b58d0d9818a775eb66468cae9d733de9d/content/docs/tutorials/acme/nginx-ingress.md?plain=1#L104-L105).

(Which produces [this page](https://cert-manager.io/docs/tutorials/acme/nginx-ingress/#step-4---deploy-an-example-service))

That would ensure that the YAML in this document doesn't get out of sync with the YAML in certificate.yaml.

I like that you've got an explicit link to it as well as embedding it, in any case!

wallrj replied:
Done. In https://github.com/cert-manager/website/pull/1140

SgtCoDFish commented on content/docs/tutorials/getting-started-aks-letsencrypt/README.md:
````diff
@@ -0,0 +1,812 @@
+```
+🔗 <a href="values.yaml">`values.yaml`</a>
+
+```
+helm upgrade cert-manager jetstack/cert-manager \
+    --namespace cert-manager \
+    --reuse-values \
+    --values values.yaml
````
nit (non-blocking): missing a tag for the code block, but it doesn't really matter!

wallrj replied:
Done https://github.com/cert-manager/website/pull/1140

SgtCoDFish reviewed (approved):
/lgtm
/approve
/hold

I've added a couple of suggestions - should've thought about the embedded file thing before. Happy to merge as-is, but added a hold in case you'd like to embed the files!

SgtCoDFish commented:
Should also say that I've not tested this (no Azure env) but I think it looks _awesome_!

wallrj commented:
/hold cancel
